### PR TITLE
Add a job to prepare ioda format dumps for use in atmospheric UFS-DA

### LINKED ISSUE #1820: Add job to prepare ioda format dumps for UFS-DA
**Description**
This  issue is opened to document the addition of a g-w job to convert operational bufr dump files to the ioda format used by UFS-DA applications.

**Requirements**
Create a new j-job to convert bufr dumps to ioda format.

**Acceptance Criteria (Definition of Done)**
The end goal is to run a cycled UFS-DA parallel from the bufr GDA

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
# Description
Add a job to convert bufr format GDA dumps to ioda format for use in atmospheric UFS-DA.

Resolves #1820
Depends on 
- [x] g-w #1829 
- [x] GDASAapp [#601](https://github.com/NOAA-EMC/GDASApp/pull/601) - _replace run_bufr2ioda.sh with run_bufr2ioda.py_

# Type of change
- New feature (adds functionality)

This PR adds j-job `JGLOBAL_PREP_IODA_OBS` and  rocoto driver `prepiodaobs.sh` to the suite of cycled gfs jobs.  Changes are made in workflow scripts to generate the xml required to run prepioodaobs.  The new job is added to `config.resources` along with the addition of a new config file, `config.prepiodaobs`.  

At present [GDASApp](https://github.com/NOAA-EMC/GDASApp) only converts bufr satwnd dumps to ioda format.  Additional bufr dump types will be added in the future.

# Change characteristics
- Is this a breaking change (a change in existing functionality)? **_NO_**
- Does this change require a documentation update? **_NO_**

# How has this been tested?
- Clone and build on Hera and Orion
- Cycled test on Hera  Run C384/C192 L127 for 2021081418.

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: jobs/rocoto/prepiodaobs.sh
Comment (WalterKolczynski-NOAA) [inline jobs/rocoto/prepiodaobs.sh]: Please remove the dead code.

File: parm/config/gfs/config.resources
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.resources]: I know this script is inconsistent in general, including the `preplandobs` directly above it that uses this identical pattern, but can we at least make this new section use a consistent style? Two variables are defined then exported, another is exported as it is defined.

**Review by WalterKolczynski-NOAA — COMMENTED** [review]
I think we need more information about what the exscript is actually doing, especially with the `$COM_OBS` directory. In operations, that directory does NOT fall under GFS (it is an obsproc directory). If the script is placing output there (and I don't see any other COM directories being defined), that is probably a no-go. It might be okay temporarily for development, but we need to start thinking about making all this new JEDI stuff ops-ready.

File: jobs/rocoto/prepiodaobs.sh
Comment (RussTreadon-NOAA) [reply jobs/rocoto/prepiodaobs.sh]: Good catch.  My bad.  Unused lines removed.  Done at 3d614c23

File: parm/config/gfs/config.resources
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.resources]: True, the `prepiodaobs` section is a direct copy of the preceding `preplandobs` section with _ioda_ replacing _land_.  I assumed the `preplandobs` section was OK since it entered `develop` via PR #1609.   

Refactor `prepiodaobs` section following `waveprep` section.  Done at 3d614c23

**Comment by RussTreadon-NOAA** [comment]
> I think we need more information about what the exscript is actually doing, especially with the `$COM_OBS` directory. In operations, that directory does NOT fall under GFS (it is an obsproc directory). If the script is placing output there (and I don't see any other COM directories being defined), that is probably a no-go. It might be okay temporarily for development, but we need to start thinking about making all this new JEDI stuff ops-ready.

A [comment](https://github.com/NOAA-EMC/global-workflow/issues/1820#issuecomment-1700791710) was added to issue #1820 briefly describing the exscript.  GDASApp PR [#575](https://github.com/NOAA-EMC/GDASApp/pull/575)  provides additional details regarding bufr2ioda processing.   

I believe you are correct.  The bufr to ioda processing in this PR is a temporary step for UFS-DA prototyping.  Please contact @CoryMartin-NOAA, @emilyhcliu, and @ilianagenkova for long term (ie, operational) plans for processing observations using UFS-DA software. 

Is the above sufficient for the time being?

You make a valid point, @WalterKolczynski-NOAA , regarding the relationship between ObsProc and GFS.  g-w has long exercised pieces of ObsProc via `prep.sh`.  Now we are adding more observation processing stuff which could / should reside in ObsProc.  That said, the [ObsProc](https://github.com/NOAA-EMC/obsproc) repo does not maintain a workflow.  Therefore, we use g-w to run pieces of ObsProc.  What's the best long term solution?

File: jobs/JGLOBAL_PREP_IODA_OBS
Comment (aerorahul) [inline jobs/JGLOBAL_PREP_IODA_OBS]: I don't see this script in `HOMEgfs`.

File: parm/config/gfs/config.prepiodaobs
Comment (aerorahul) [inline parm/config/gfs/config.prepiodaobs]: This `ush` script should be in `global-workflow` not `GDASApp`.

File: parm/config/gfs/config.prepiodaobs
Comment (aerorahul) [inline parm/config/gfs/config.prepiodaobs]: Should be in `global-workflow` or soft-linked to `parm` in `global-workflow` via `link_workflow.sh`.

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: I would suggest calling this `prepatmiodaobs` to identify the `atm` component of observation processing.  Marine obs will need ioda preparation jobs and it should be distinguishable.

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: update `prepiodaobs` based on previous comment.

File: parm/config/gfs/config.prepiodaobs
Comment (aerorahul) [reply parm/config/gfs/config.prepiodaobs]: This should be in python to align with all new scripts that are being added in global-workflow.

**Review by aerorahul — CHANGES_REQUESTED** [review]
I have several questions on this PR:
- see inline specific comments
- how does the ioda processing for atmosphere observations in general fit in the scope of the global-workflow?  without knowing that the scope or expanse is, it is hard to provide robust review that may also overlap with similar activities that fall under "prep ioda observations for JEDI applications"

**Comment by RussTreadon-NOAA** [comment]
Thank you @aerorahul for your feedback.   @CoryMartin-NOAA and @emilyhcliu will need to reply to some of your questions.  I took their development and added hooks to exercise it from g-w.  You ask excellent design questions which need to go back to source developers.

**Comment by CoryMartin-NOAA** [comment]
@aerorahul as previously discussed, much of what is here is intended as a stopgap for prototype development and not intended for the final operational implementation. If you would prefer that we take all observation processing out of the workflow and just have the workflow handle ln/cp, we can do so.

**Comment by aerorahul** [comment]
> @aerorahul as previously discussed, much of what is here is intended as a stopgap for prototype development and not intended for the final operational implementation. If you would prefer that we take all observation processing out of the workflow and just have the workflow handle ln/cp, we can do so.

That is only partially true @CoryMartin-NOAA 
In the case of marine observations, pre-processing is desired to be inline with the workflow.
Even in the case of atmosphere observations, I was told there will be some aspect of preprocessing of observations that will be relegated to the GFS application, so while this particular implementation of preprocessing is a stopgap for prototype development, the idea is not.  
Implementations can change under the hood.

File: jobs/JGLOBAL_PREP_IODA_OBS
Comment (RussTreadon-NOAA) [reply jobs/JGLOBAL_PREP_IODA_OBS]: Good catch.  I set `BUFR2IODASH` in `config.prepiodaobs` and failed to see this error.   `BUFR2IODASH` currently resides in GDASApp `ush/ioda/bufr2ioda`   I added a soft link to g-w `parm` via `link_workflow.sh`.  The path in `JGLOBAL_PREP_IODA_OBS` was changed to `${HOMEgfs}/ush/run_bufr2ioda.sh`.   Done at 7af58ee3.

File: parm/config/gfs/config.prepiodaobs
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.prepiodaobs]: The path question has been addressed via a soft link from GDASApp to g-w.   See comment above.   Done at 7af58ee3.

DAD developers can write a python equivalent of `run_bufr2ioda.sh` in GDASApp.  Would it be acceptable to do this via a separate PR - one which updates the GDASApp hash in `Externals.cfg` and `sorc/checkout.sh`?

File: parm/config/gfs/config.prepiodaobs
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.prepiodaobs]: Soft link added at 7af58ee3.

File: parm/config/gfs/config.resources
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.resources]: Replace `prepiodaobs` with `prepatmiodaobs`.   Done at 7af58ee3.

File: parm/config/gfs/config.resources
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.resources]: `prepiodaobs` replaced with `prepatmiodaobs`.   Done at 7af58ee3.

**Comment by RussTreadon-NOAA** [comment]
Rerun 20210814 18 gdas `prepatmiodaobs` on Hera using `RussTreadon-NOAA:feature/iodaobs` at cd6972d4.   Job successfully ran to completion.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

File: parm/config/gfs/config.prepiodaobs
Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.prepiodaobs]: Open GDASApp issue [#599](https://github.com/NOAA-EMC/GDASApp/issues/599) to document creation and testing of `run_bufr2ioda.py`.   Python script tested and confirmed to generate identical output with respect to bash script.

One detail to be resolved is the g-w use of variable `CDUMP` to identify the current dump versus bufr2ioda use of variable `DUMP` to identify the dump being processed.   For testing purposes modify working copy of `JGLOBAL_ATM_PREP_IODA_OBS` to read
```
# Run relevant script
 EXSCRIPT=${BUFR2IODASH:-${HOMEgfs}/ush/run_bufr2ioda.py}
 export DUMP=$CDUMP
 ${EXSCRIPT} "${PDY}${cyc}" "${CDUMP}" "${DMPDIR}" "${IODAPARM}" "${COM_OBS}/"
```
Need to discuss with EID and DAD developers how to address this detail.

Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.prepiodaobs]: Tested an alternative solution to the above issue.  Remove `export DUMP=$CDUMP` from `JGLOBAL_ATM_PREP_IODA_OBS` and instead replace `DUMP` with `CDUMP` in `gdas.cd/parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_goes.json`.   That is, make the following change
```
   "data_type"        : "satwnd",
-  "cycle_type"       : "{{ DUMP }}",
+  "cycle_type"       : "{{ CDUMP }}",
   "cycle_datetime"   : "{{ current_cycle | to_YMDH }}",
```
in `bufr2ioda_satwind_amv_goes.json`.

Rerun 20210814 18 gdas prepatmiodaobs.  Job successfully ran to completion.

Two modifications have been tested and both result in successful runs.  The question for EID and DAD developers is which approach to use?
1. modify g-w:  `export DUMP=$CDUMP` from `JGLOBAL_ATM_PREP_IODA_OBS`
2. modify GDASAPP:  replace `DUMP` with `CDUMP` in  `bufr2ioda_satwind_amv_goes.json`
3. another approach developed by the team

Tag @CoryMartin-NOAA , @emilyhcliu , and @aerorahul

Comment (RussTreadon-NOAA) [reply parm/config/gfs/config.prepiodaobs]: Given that g-w primarily uses `CDUMP` for the current cycle, decide to replace variable `DUMP` in GDASApp `parm/ioda/bufr2ioda/bufr2ioda_satwind_amv_goes.json` with `CDUMP`.  Done in GDASApp branch `feature/bufr2iodapy` at [`64d330a`](https://github.com/NOAA-EMC/GDASApp/commit/64d330a740f8c41af7e26dd2e3baced74072bf05).  

GDASApp PR [#601](https://github.com/NOAA-EMC/GDASApp/pull/601) opened to merge `feature/bufr2iodapy` into GDASApp `develop`.   Once GDASApp PR [#601](https://github.com/NOAA-EMC/GDASApp/pull/601) is approved and merged into `develop`, update `Externals.cfg` and `sorc/checkout.sh` to the appropriate GDASApp hash.

**Comment by RussTreadon-NOAA** [comment]
**Note:**  Once GDASApp PR #601 is merged into GDASAPP `develop` and the updated `develop` is pulled into GDASApp `feature/stable-nightly`, we need to update the GDASApp hash in `RussTreadon-NOAA:feature/iodaobs` `Externals.cfg` and `sorc/checkout.sh`.

**Comment by RussTreadon-NOAA** [comment]
Test GDASApp branch `feature/bufr2iodapy` along with g-w branch `RussTreadon-NOAA:feature/iodaobs` on Hera and Orion for 20210814 18 gdas.   Job `prepatmiodaobs` ran to completion on both machines.  IODA dump files were created for GOES-16 and GOES-17 ABI from the bufr satwnd dump file.

The last update I am aware of for this PR is to update `Externals.cfg` and `sorc/checkout.sh` with an updated GDASApp hash once GDASApp PR [#601](https://github.com/NOAA-EMC/GDASApp/pull/601) is merged into develop and `feature/stable-nightly` is updated.

Does this PR require any additional revisions besides the updated GDASApp hash?

**Comment by RussTreadon-NOAA** [comment]
GDASApp PR [#601](https://github.com/NOAA-EMC/GDASApp/pull/601) was merged into GDASApp `develop` at [`0073a5f`](https://github.com/NOAA-EMC/GDASApp/commit/0073a5fdc96817fa96e0dc3469a5650e17ca229a).   This update was pulled into GDASApp branch [`feature/stable-nightly`](url) at [`ac8fdb1`](https://github.com/NOAA-EMC/GDASApp/commit/ac8fdb17eb297bc2133b551eab494f008bc29b4c).

g-w points at GDASApp `feature/stable-nightly`.   Update the GDASApp hash to [`ac8fdb1`](https://github.com/NOAA-EMC/GDASApp/commit/ac8fdb17eb297bc2133b551eab494f008bc29b4c) in `RussTreadon-NOAA:feature/iodaobs` `Externals.cfg` and `sorc/checkout.sh`.

Done at 339c6a02.

**Comment by RussTreadon-NOAA** [comment]
**Hera and Orion tests**
Install `RussTreadon-NOAA:feature/iodaobs` at 339c6a02 on both machines.  Confirm that GDASApp checkout is at the correct hash.  Run 20210814 18 gdas `prepatmiodaobs`.   Job successfully ran to completion on both machines.

Work for this PR is done apart from possible requests from reviewers.   Re-reviews requested.

@WalterKolczynski-NOAA , do I need to trigger any of the Orion or Hera CI tests using labels?

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA , do I need to trigger any of the Orion or Hera CI tests using labels?

This is an action we reserve for the g-w code managers. We'll take care of that if/when necessary.

**Comment by RussTreadon-NOAA** [comment]
> > @WalterKolczynski-NOAA , do I need to trigger any of the Orion or Hera CI tests using labels?
> 
> This is an action we reserve for the g-w code managers. We'll take care of that if/when necessary.

Got it.  Thank you @WalterKolczynski-NOAA for the clarification.   

Is there anything you or @aerorahul need me to do for this PR?   I'm done with the changes I intend to make.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Looks fine to me. Until CI exercises the UFS-DA jobs, I don't think there is much point in CI testing it.

**Comment by aerorahul** [comment]
The changes look good with the exception that all scripts should be part of the global-workflow if they are executed in the global applications.  This has been established for a while now. 
This PR adds scripts that are not part of the global-workflow and therefore are hard to manage and maintain a conformity across the project.
We can discuss this offline, but imagine if all the various components started writing and managing their own configuration, scripting, etc, we would be back to where we were.

**Comment by WalterKolczynski-NOAA** [comment]
~~Actually, we do have a test for these now in CI, don't we?~~

**Comment by RussTreadon-NOAA** [comment]
> The changes look good with the exception that all scripts should be part of the global-workflow if they are executed in the global applications. This has been established for a while now. This PR adds scripts that are not part of the global-workflow and therefore are hard to manage and maintain a conformity across the project. We can discuss this offline, but imagine if all the various components started writing and managing their own configuration, scripting, etc, we would be back to where we were.

Please advise as to the proper action to take to move this PR forward.   Tagging @CoryMartin-NOAA, @emilyhcliu , and @guillaumevernieres for awareness.

**Review by aerorahul — APPROVED** [review]

**Comment by aerorahul** [comment]
@RussTreadon-NOAA 
I didn't mean to hold this PR.  I was simply stating what has been a precedent for a while.  An issue has been opened now specifically for UFS-DA https://github.com/NOAA-EMC/global-workflow/issues/1839 and https://github.com/NOAA-EMC/global-workflow/issues/1840.  Discussion can be moved/continued there.

**Comment by RussTreadon-NOAA** [comment]
Thank you @aerorahul for approving this PR.   Thank you @WalterKolczynski-NOAA for merging this PR into `develop`.

**Comment by ilianagenkova** [comment]
My input probably won't matter much by now, but I think both Cory and Rahul
are right. Some observations pre-processing might fall in the g-w, but it's
not been discussed in detail. Specifically for marine data, some decisions
are still to be made.  Perhaps a meeting at some point would answer some
questions.

**Pull request merged**
